src/runtime_src/driver/zynq/drm/zocl/sched_exec.c

```diff
@@ -454,6 +477,105 @@ static irqreturn_t sched_exec_isr(int irq, void *arg)
 	return IRQ_HANDLED;
 }
 
+static void
+init_cu_by_idx(struct sched_cmd *cmd, int cu_idx)
+{
+	u32 i, size = regmap_size(cmd);
+	u32 *virt_addr = cu_idx_to_addr(cmd->ddev, cu_idx);
+	struct ert_init_kernel_cmd *ik;
+
+	ik = (struct ert_init_kernel_cmd *)cmd->packet;
+
+	/* Initialize CU by writing register map starting at
+	 * base_address + 0x10 (byte)
+	 * This based on the fact that kernel used
+	 *	0x00 -- Control Register
+	 *	0x04 and 0x08 -- Interrupt Enable Registers
+	 *	0x0C -- Interrupt Status Register
+	 * Skip the first 4 words in user regmap.
+	 */
+	for (i = 4; i < size; ++i)
+		iowrite32(*(ik->data + ik->extra_cu_masks + i), virt_addr + i);
+}
+
+/**
+ * init_cus() - Initialize CUs from user space command.
+ *
+ * Process the initialize CUs command sent from user space. Only one process
+ * can initialize a given CU, so if a CU in the CU masks is initialized,
+ * this function ignores the initialize request.
+ *
+ * The initialization is done by copying the user regmap to register map.
+ */
+static void
+init_cus(struct sched_cmd *cmd)
+{
+	struct drm_zocl_dev *zdev = cmd->ddev->dev_private;
+	struct ert_init_kernel_cmd *ik;
+	unsigned int i;
+	uint32_t *cmp;
+	int num_masks = cu_masks(cmd);
+	int mask_idx;
+	int warn_flag = 0;
+
+	ik = (struct ert_init_kernel_cmd*)cmd->packet;
+	cmp = &ik->cu_mask;
+
+	for (mask_idx = 0; mask_idx < num_masks; ++mask_idx) {
+		u32 cmd_mask = cmp[mask_idx];
+		u32 inited_mask = zdev->exec->cu_init[mask_idx];
+		u32 uninited_mask = (cmd_mask | inited_mask) ^ inited_mask;
+		u32 busy_mask = zdev->exec->cu_status[mask_idx];
+
+		/*
+		 * If some requested CUs are initialized already,
+		 * record a flag and post some log later.
+		 */
+		if (!warn_flag && (inited_mask & cmd_mask))
+			warn_flag = 1;
+
+		/*
+		 * We don't have uninitialized CU for this 32 mask_id.
+		 * Move to next 32 CUs.
+		 */
+		if (!uninited_mask)
+			continue;
+
+		for (i = 0; i < 32; ++i) {
+			int cu_idx;
+
+			if (!(uninited_mask & (1 << i))) 
```

Comment:
Not sure whether "ffs_or_neg_one(u32 mask)" is more efficiency than check each bit in a mask. Pointing this out for more thought. I'm good with current implementation.

Reply:
This should not matter that much since this cu_init is done just once. The reason I did not choose ffs is because we need an extra step to clear the bit we found in uninited_mask by ffs.